### ISSUE TITLE
Add Streams.iterate(T seed, UnaryOperator<T> f, Predicate<T> stop)

### ISSUE THREAD

**Issue by Iteratrix** [google/guava#3252]
Stream has a two argument version of `iterate` that takes a seed (`T seed`) and an iterative function (`UnaryOperator<T> f`) and generates an infinite stream by repeatedly applying the function. This is an almost perfect replacement for a traditional for loop, except that it's missing a termination predicate. Java 9 adds `takeWhile` which can be used to terminate a stream generated this way, but unfortunately some of us aren't beyond Java 8 yet.

It might be nice to add a static method `<T> Stream<T> iterate(T seed, UnaryOperator<T> f, Predicate<T> stop)` to the `Streams` utility class that could fill this niche, at least until Java 11 is out and widely considered stable. An alternative (or possibly additional) approach would be adding a static `takeWhile` method to the same class that would limit the Stream by some predicate, though that's a bit clunkier.

Thanks, and let me know if you need more details or any clarification.

**Comment by fluentfuture** [comment]
Or, a flatMap variant that supports both fanning out and fanning in (termination)?

```java
static Stream<T> generate(T initial, Function<T, Stream<T>> iterate);
```

With this, ```iterate(seed, f, stop)``` can be expressed as:
```java
generate(seed, v -> stop(v) ? empty() : f.apply(v));
```

**Comment by Diagoras** [comment]
@fluentfuture True, that would be more generic. I guess the question is whether Guava wants to just address this use case or provide something more general. Also, won't stack overflow be a problem with that method, since it can't be implemented iteratively?

**Comment by fluentfuture** [comment]
What's the use case you have in mind for this particular case, @Diagoras ?

Personally it feels that ```iterate(initial, step, stop)``` is too similar to ```for (T v = initial; !stop(v); v = step(v)) {...}```, that it may beg the question whether introducing a functional dialect to the plain old for loop pulls the weight.

Though I guess ```forEach()``` is a precedent despite having ```for (T v : elements)```.

Implementation-wise, the following ```generate()``` should work in Java 10 (or maybe 9+?) without stack overflow (perhaps not the most efficient):

```java
Stream<T> generate(T initial, Function<T, Stream<T>> step) {
  return concat(Stream.of(initial), step.apply(initial).flatMap(n -> generate(n, step)));
}
```

**Comment by Diagoras** [comment]
@fluentfuture It's pretty much what you've stated. I often want a for loop style generator, but with the flexibility and modern approach of a `Stream`. I can easily have that on Java 9+ (`Stream.iterate(() -> initial, previous -> step(previous)).takeWhile(!stop)`) but that syntax isn't as clear about what it's doing and is unavailable before Java 9.

Since Guava has a number of helpful methods for pre-Java 9 streaming in the `Streams` class (for example, `<T> stream(Optional<T> optional)` which is superseded by `Optional#toStream()` in Java 9) I figured this method would fit right in. If you think the use case is too niche, though, then we can close this issue instead.

**Comment by fluentfuture** [comment]
Speculating about the API, I think the ```iterate(initial, step, stop)``` signature as is may not be self-evident enough. It's equally likely to think of the predicate as ```while``` instead of ```until``` because it's what a for-loop will look like.

And I suppose adding two similar APIs like ```iterateUntil``` and ```iterateWhile``` might be too much for little gain.

On the other hand, the more generic ```generate(initial, iterate)``` signature could be used to create a less ambiguous syntax, but you'll need to add two helpers ```streamWhile()``` and ```streamUntil()``` in the *local code base*.

```java
Function<T, Stream<T>> streamUntil(Predicate<T> condition, UnaryOperator<T> step) {
  return v -> condition.test(v) ? Stream.of(step.apply(v)) : Stream.empty();
}
Function<T, Stream<T>> streamWhile(Predicate<T> condition, UnaryOperator<T> step) {
  return streamUntil(condition.negate(), step);
}
```

Then:

```java
generate(initial, streamUntil(this::shouldStop, this::nextValue));
generate(initial, streamWhile(this::shouldContinue, this::nextValue));
```
